app/controllers/supervision/AnotherBodyController.scala

```diff
@@ -29,59 +29,66 @@ import views.html.supervision.another_body
+            val form: Form2[AnotherBody] = (for {
+              supervision <- response
+              anotherBody <- supervision.anotherBody
+            } yield Form2[AnotherBody](anotherBody)).getOrElse(EmptyForm)
+            Ok(another_body(form, edit))
+        }
   }
 
-  def post(edit : Boolean = false) = Authorised.async {
-    implicit authContext => implicit request =>
-      Form2[AnotherBody](request.body) match {
-        case f: InvalidForm =>
-          Future.successful(BadRequest(another_body(f, edit)))
-        case ValidForm(_, data: AnotherBody) =>
-          for {
-            supervision <- dataCacheConnector.fetch[Supervision](Supervision.key)
-            cache <- dataCacheConnector.save[Supervision](Supervision.key,
-              updateData(supervision, data))
-          } yield {
-            redirectTo(edit, cache)
-          }
-      }
+  def post(edit: Boolean = false) = Authorised.async {
+    implicit authContext =>
+      implicit request =>
+        Form2[AnotherBody](request.body) match {
+          case f: InvalidForm =>
+            Future.successful(BadRequest(another_body(f, edit)))
+          case ValidForm(_, data: AnotherBody) =>
+            for {
+              supervision <- dataCacheConnector.fetch[Supervision](Supervision.key)
+              _ <- dataCacheConnector.save[Supervision](Supervision.key, updateData(supervision, data))
+              maybeCache <- dataCacheConnector.fetchAll
+              cache <- Future.successful(maybeCache)
+            } yield redirectTo(edit, cache)
+        }
   }
 
   private def updateData(supervision: Supervision, data: AnotherBody): Supervision = {
     def updatedAnotherBody = (supervision.anotherBody, data) match {
       case (_, d) if d.equals(AnotherBodyNo) => AnotherBodyNo
-      case (Some(ab), d:AnotherBodyYes) if ab.equals(AnotherBodyNo) => AnotherBodyYes(d.supervisorName, None, None, None)
-      case (Some(ab), d:AnotherBodyYes) => ab.asInstanceOf[AnotherBodyYes].supervisorName(d.supervisorName)
-      case (None, d:AnotherBodyYes) => AnotherBodyYes(d.supervisorName, None, None, None)
+      case (Some(ab), d: AnotherBodyYes) if ab.equals(AnotherBodyNo) => AnotherBodyYes(d.supervisorName, None, None, None)
+      case (Some(ab), d: AnotherBodyYes) => ab.asInstanceOf[AnotherBodyYes].supervisorName(d.supervisorName)
+      case (None, d: AnotherBodyYes) => AnotherBodyYes(d.supervisorName, None, None, None)
 
     }
+
     supervision.anotherBody(updatedAnotherBody).copy(hasAccepted = true)
   }
 
-  private def redirectTo(edit: Boolean, cache: CacheMap)(implicit authContext: AuthContext, headerCarrier: HeaderCarrier) = {
+  private def redirectTo(edit: Boolean, maybeCache: Option[CacheMap])(implicit authContext: AuthContext, headerCarrier: HeaderCarrier) = {
 
     import utils.ControllerHelper.{anotherBodyComplete, isAnotherBodyYes, supervisionComplete}
 
-    val anotherBody = anotherBodyComplete(cache)
+    maybeCache match {
```

Comment:
If we implement this in the controllers as mentioned above and pass the cache in do we need to do this here?